Add 1.10 Release announcement. (#301)
apache/flink-web: _posts/2020-02-03-release-1.10.0.md

{% toc %}

The binary distribution and source artifacts are now available on the updated [Downloads page]({{ site.baseurl }}/downloads.html) of the Flink website. For more details, check the complete [release changelog]() and the [updated documentation]({{ site.DOCS_BASE_URL }}flink-docs-release-1.10/). We encourage you to download the release and share your feedback with the community through the [Flink mailing lists](https://flink.apache.org/community.html#mailing-lists) or [JIRA](https://issues.apache.org/jira/projects/FLINK/summary).

## New Features and Improvements

### Improved Memory Management and Configuration

The current `TaskExecutor` memory configuration in Flink has some shortcomings that make it hard to reason about or optimize resource utilization, such as: 

* Different configuration models for memory footprint in Streaming and Batch execution; 

* Complex and user-dependent configuration of off-heap state backends (i.e. RocksDB) in Streaming execution.

> To make memory options more explicit and intuitive to users, Flink 1.10 introduced significant changes to the `TaskExecutor` memory model and configuration logic ([FLIP-49](https://cwiki.apache.org/confluence/display/FLINK/FLIP-49%3A+Unified+Memory+Configuration+for+TaskExecutors)). These changes make Flink more adaptable to all kinds of deployment environments (e.g. Kubernetes, Yarn, Mesos), giving users strict control over its memory consumption.

Reviewer: ```suggestion
To make memory options more explicit and intuitive to users, Flink 1.10 introduces significant changes to the `TaskExecutor` memory model and configuration logic ([FLIP-49](https://cwiki.apache.org/confluence/display/FLINK/FLIP-49%3A+Unified+Memory+Configuration+for+TaskExecutors)). These changes make Flink more adaptable to all kinds of deployment environments (e.g. Kubernetes, Yarn, Mesos), giving users strict control over its memory consumption.
```

**Out-of-the-box Managed Memory Configuration**

Managed memory was extended to also account for memory usage of state backends. This means that there may be scenarios where jobs in the same cluster require different types of memory (e.g. a streaming job with `MemoryStateBackend` and another streaming job with `RocksDBStateBackend`). To enable these scenarios and allow users to switch between Streaming and Batch execution without having to modify cluster configurations, managed memory is now always off-heap.

**Simplified RocksDB Configuration**

> Configuring an off-heap state backend like RocksDB used to involve a good deal of manual tuning, like decreasing the JVM heap size or setting Flink to use off-heap memory. This can now be achieved through Flink's out-of-box configuration, and adjusting the memory budget for `RocksDBStateBackend` is as simple as resizing the managed memory size. 

Reviewer: ```suggestion
Configuring an off-heap state backend like RocksDB used to involve a good deal of manual tuning, like decreasing the JVM heap size or setting Flink to use off-heap memory. This can now be achieved through Flink's out-of-box configuration, making adjusting the memory budget for `RocksDBStateBackend` is as simple as resizing the managed memory size. 
```

	</center>
</span>

As introduced in [Unified Logic For Job Submission](#unified-logic-for-job-submission), all command-line options in Flink 1.10 are mapped to a unified configuration. For this reason, users can simply refer to the Kubernetes config options and submit a job to an existing Flink session on Kubernetes in the CLI using:

```bash
./bin/flink run -d -e kubernetes-session -Dkubernetes.cluster-id=<ClusterId> examples/streaming/WindowJoin.jar
```

If you want to try out this preview feature, we encourage you to walk through the [Native Kubernetes setup]({{ site.DOCS_BASE_URL }}flink-docs-release-1.10/ops/deployment/native_kubernetes.html), play around with it and share feedback with the community.

### Table API/SQL: Production-ready Hive Integration

Hive integration was announced as a preview feature in Flink 1.9. This first implementation allowed users to persist Flink-specific metadata in Hive Metastore using SQL DDL, call UDFs defined in Hive and use Flink for reading and writing Hive tables. However, some further development was still required to really bring production-ready Hive integration to Flink.

> 

Reviewer 2: ```
#### Compatibility with More Hive Versions

Flink 1.9 only natively supports Hive 1.2.1 and 2.3.4, and relies on Hive's own compatibility for the rest of Hive 1.x.x and 2.x.x versions.

Flink 1.10 introduces [native compatibility for all the Hive 1.x.x, 2.x.x, and 3.x.x versions]({{ site.DOCS_BASE_URL }}flink-docs-release-1.10/dev/table/hive/#supported-hive-versions).

#### Enhanced Experience to Persist Flink Tables in Hive Metastore

In Flink 1.9, we introduced a new beta feature to allow users persist their Flink tables created from DDL into Hive Metastore. It solves users' pain point of unable to reuse Flink tables created from DDL across sessions and having to re-create those tables every time.

We made the feature more robust and the experience more pleasant in Flink 1.10 and made it compatible with all 1.x.x, 2.x.x, and 3.x.x Hive Metastore versions.
```

Author: How relevant would you consider this in comparison to what is already covered for Hive? It seems really prominent in the first few paragraphs of entering the Hive Integration documentation. 

I'm afraid the blogpost is getting too exhaustive/extensive and would prefer to leave this kind of details to a follow-up blogpost dedicated to the Flink-Hive integration. What do you think?

Reviewer 2: Sounds reasonable. I'll incorporate this into the 1st paragraph

Author: Thanks, @bowenli86! @MarkSfik will get in touch with you about the blogpost.

* **Projection pushdown:** Flink leverages projection pushdown to minimize data transfer between Flink and Hive tables by omitting unnecessary fields from table scans. It is especially beneficial when a table contains many columns.

* **LIMIT pushdown:** For queries with the `LIMIT` clause, Flink will limit the number of output records wherever possible to minimize the amount of data transferred across the network.

* **ORC Vectorization on Read**

#### Pluggable Modules as Flink System Objects (Beta)

Flink 1.10 introduces a generic mechanism for pluggable modules in the Flink table core, with a first focus on system functions ([FLIP-68](https://cwiki.apache.org/confluence/display/FLINK/FLIP-68%3A+Extend+Core+Table+System+with+Pluggable+Modules)). With modules, users can extend Flink’s system objects — for example use Hive built-in functions that behave like Flink system functions. This release ships with a pre-implemented `HiveModule`, supporting multiple Hive versions, but users are also given the possibility to [write their own pluggable modules]({{ site.DOCS_BASE_URL }}flink-docs-release-1.10/dev/table/modules.html).

> ### Other Improvements to the Table API/SQL

Author: Initially added this title to be consistent with the structure of previous blogposts. Also, one of the points is about TPC-DS support, which doesn't really fall under that category.

#### Watermarks and Computed Columns in SQL DDL

Flink 1.10 supports stream-specific syntax extensions to define time attributes and watermark generation in Flink SQL DDL ([FLIP-66](https://cwiki.apache.org/confluence/display/FLINK/FLIP-66%3A+Support+Time+Attribute+in+SQL+DDL)). This allows time-based operations, like windowing, and the definition of [watermark strategies]({{ site.DOCS_BASE_URL }}flink-docs-release-1.10/dev/table/sql/create.html#create-table) on tables created using DDL statements.

```sql
CREATE TABLE table_name (

  WATERMARK FOR columnName AS <watermark_strategy_expression>

) WITH (
  ...
)
```

This release also introduces support for virtual computed columns ([FLIP-70](https://cwiki.apache.org/confluence/display/FLINK/FLIP-70%3A+Flink+SQL+Computed+Column+Design)) that can be derived based on other columns in the same table or deterministic expressions (i.e. literal values, UDFs and built-in functions).

> 

Author: Sure. I'll also add the other DDL extensions from FLIP-69.

#### Additional Extensions to SQL DDL

There is now a clear distinction between temporary/persistent and system/catalog functions ([FLIP-57](https://cwiki.apache.org/confluence/display/FLINK/FLIP-57%3A+Rework+FunctionCatalog)). This not only eliminates ambiguity in function reference, but also allows for deterministic function resolution order (i.e. in case of naming collision, system functions will precede catalog functions, with temporary functions taking precedence over persistent functions for both dimensions).

> In order to correctly handle and guarantee a consistent behavior across meta-object (tables, views, functions) in the future, some object declaration methods in the Table API  have been deprecated in favor of methods that are closer to standard SQL DDL ([FLIP-64](https://cwiki.apache.org/confluence/display/FLINK/FLIP-64%3A+Support+for+Temporary+Objects+in+Table+module)).

Reviewer: ```suggestion
In order to correctly handle and guarantee a consistent behavior across meta-objects (tables, views, functions) in the future, some object declaration methods in the Table API  have been deprecated in favor of methods that are closer to standard SQL DDL ([FLIP-64](https://cwiki.apache.org/confluence/display/FLINK/FLIP-64%3A+Support+for+Temporary+Objects+in+Table+module)).
```

---
layout: post 
title:  "Apache Flink 1.10.0 Release Announcement" 

> date: 2020-02-03T02:30:00.000Z

Reviewer 3: The date will need to be adjusted accordingly, as well as the name of the file.

categories: news
authors:
- morsapaes:
  name: "Marta Paes"
  twitter: "morsapaes"
---

The Apache Flink community is excited to hit the double digits and announce the release of Flink 1.10.0! As a result of the biggest community effort to date, with over 1.2k issues implemented and 218 contributors, this release introduces significant improvements to the overall performance and stability of Flink jobs, a preview of native Kubernetes integration and great advances in Python support (PyFlink). 

Flink 1.10 also marks the completion of the [Blink integration](https://flink.apache.org/news/2019/08/22/release-1.9.0.html#preview-of-the-new-blink-sql-query-processor), hardening streaming SQL and bringing mature batch processing to Flink with production-ready Hive integration and TPC-DS coverage. This blog post describes all major new features and improvements, important changes to be aware of and what to expect moving forward.

{% toc %}

> The binary distribution and source artifacts are now available on the updated [Downloads page]({{ site.baseurl }}/downloads.html) of the Flink website. For more details, check the complete [release changelog]() and the [updated documentation]({{ site.DOCS_BASE_URL }}flink-docs-release-1.10/). We encourage you to download the release and share your feedback with the community through the [Flink mailing lists](https://flink.apache.org/community.html#mailing-lists) or [JIRA](https://issues.apache.org/jira/projects/FLINK/summary).

Reviewer 4: @morsapaes It seems there's still an empty log?

Reviewer 3: https://flink.apache.org/downloads.html will be updated by #302 accordingly.

Reviewer 3: The link for `release changelog` is truly missing.

Reviewer 3: Here is the link for the changelog, FYI:
https://issues.apache.org/jira/secure/ReleaseNote.jspa?projectId=12315522&version=12345845

Author: Thanks for spotting it, @carp84 and @xintongsong!

## New Features and Improvements

### Improved Memory Management and Configuration

The current `TaskExecutor` memory configuration in Flink has some shortcomings that make it hard to reason about or optimize resource utilization, such as: 

* Different configuration models for memory footprint in Streaming and Batch execution; 

* Complex and user-dependent configuration of off-heap state backends (i.e. RocksDB) in Streaming execution.

To make memory options more explicit and intuitive to users, Flink 1.10 introduces significant changes to the `TaskExecutor` memory model and configuration logic ([FLIP-49](https://cwiki.apache.org/confluence/display/FLINK/FLIP-49%3A+Unified+Memory+Configuration+for+TaskExecutors)). These changes make Flink more adaptable to all kinds of deployment environments (e.g. Kubernetes, Yarn, Mesos), giving users strict control over its memory consumption.

**Out-of-the-box Managed Memory Configuration**

> Managed memory was extended to also account for memory usage of state backends. This means that there may be scenarios where jobs in the same cluster require different types of memory (e.g. a streaming job with `MemoryStateBackend` and another streaming job with `RocksDBStateBackend`). To enable these scenarios and allow users to switch between Streaming and Batch execution without having to modify cluster configurations, managed memory is now always off-heap.

Author: Thanks for the insight and the suggestion, Tony, that was not my understanding of the changes. 👍

> Managed memory was extended to also account for memory usage of state backends. This means that there may be scenarios where jobs in the same cluster require different types of memory (e.g. a streaming job with `MemoryStateBackend` and another streaming job with `RocksDBStateBackend`). To enable these scenarios and allow users to switch between Streaming and Batch execution without having to modify cluster configurations, managed memory is now always off-heap.

Reviewer 4: I guess you were reading the FLIP-49 wiki page? We discussed and decided to change some parts of the original design according to the new findings in implementation. There were some discussion threads on the dev-ML, which must be easy to overlook.
Sorry for the confusion due to outdated wiki page, I'll update it asap.

<span>
	<center>
	<img vspace="8" style="width:75%" src="{{site.baseurl}}/img/blog/2020-02-03-release-1.10.0/flink_1.10_nativek8s.png"/>
	</center>
</span>

As introduced in [Unified Logic For Job Submission](#unified-logic-for-job-submission), all command-line options in Flink 1.10 are mapped to a unified configuration. For this reason, users can simply refer to the Kubernetes config options and submit a job to an existing Flink session on Kubernetes in the CLI using:

```bash
./bin/flink run -d -e kubernetes-session -Dkubernetes.cluster-id=<ClusterId> examples/streaming/WindowJoin.jar
```

If you want to try out this preview feature, we encourage you to walk through the [Native Kubernetes setup]({{ site.DOCS_BASE_URL }}flink-docs-release-1.10/ops/deployment/native_kubernetes.html), play around with it and share feedback with the community.

### Table API/SQL: Production-ready Hive Integration

> Hive integration was announced as a preview feature in Flink 1.9. This first implementation allowed users to persist Flink-specific metadata in Hive Metastore using SQL DDL, call UDFs defined in Hive and use Flink for reading and writing Hive tables. Flink 1.10 rounds up this effort with further developments that bring production-ready Hive integration to Flink.

Reviewer 2: ```suggestion
Hive integration was announced as a preview feature in Flink 1.9. This preview allowed users to persist Flink-specific metadata (e.g. Kafka tables) in Hive Metastore using SQL DDL, call UDFs defined in Hive and use Flink for reading and writing Hive tables. Flink 1.10 rounds up this effort with further developments that bring production-ready Hive integration to Flink with full compatibility of [most Hive versions]({{ site.DOCS_BASE_URL }}flink-docs-release-1.10/dev/table/hive/#supported-hive-versions).
```

<span>
	<center>
	<img vspace="8" hspace="100" style="width:75%" src="{{site.baseurl}}/img/blog/2020-02-03-release-1.10.0/flink_1.10_pyflink.gif"/>
	</center>
</span>

If you are interested in the underlying implementation — leveraging Apache Beam’s [Portability Framework](https://beam.apache.org/roadmap/portability/) — refer to the “Architecture” section of FLIP-58 and also to [FLIP-78](https://cwiki.apache.org/confluence/display/FLINK/FLIP-78%3A+Flink+Python+UDF+Environment+and+Dependency+Management). These data structures lay the required foundation for Pandas support and for PyFlink to eventually reach the DataStream API. 

From Flink 1.10, users can also easily install PyFlink through `pip` using:

```bash
pip install apache-flink
```

For a preview of other improvements planned for PyFlink, check [FLINK-14500](https://issues.apache.org/jira/browse/FLINK-14500) and get involved in the [discussion](http://apache-flink.147419.n8.nabble.com/Re-DISCUSS-What-parts-of-the-Python-API-should-we-focus-on-next-td1285.html) for requested user features.

> ## Important Changes

Reviewer 5: Can we also add Elasticsearch 7.x support as an important change(feature). @MarkSfik

Author: Will add it in, @yanghua.
